zerver/actions/message_delete.py

```diff
@@ -52,6 +52,28 @@ def do_delete_messages(realm: Realm, messages: Iterable[Message]) -> None:
         archiving_chunk_size = retention.STREAM_MESSAGE_BATCH_SIZE
 
     move_messages_to_archive(message_ids, realm=realm, chunk_size=archiving_chunk_size)
+    if message_type == "stream": # nocoverage
+        stream = Stream.objects.get(id=sample_message.recipient.type_id)
+        if stream.first_message_id:
+            first_stream_message_exists = Message.objects.filter(
+                id=stream.first_message_id
+            ).exists()
+            if not first_stream_message_exists and stream.recipient_id:
+                stream_messages = Message.objects.filter(
+                    realm_id=realm.id, recipient_id=stream.recipient_id
+                ).order_by("date_sent")
+                if stream_messages:
+                    stream.first_message_id = stream_messages[0].id
+                    stream.save(update_fields=["first_message_id"])
```

Comment:
This is a poor algorithm. The first thing you should be checking is whether any of the deleted messages is the current `first_message_id` of the stream; if none are, then, there's no need to touch the `Message` table at all. 

You should also consider putting this logic in a function, since we likely need to call it in code paths that move messages out of streams, too. Doing so also makes it easy to use the early return pattern to avoid the extreme nesting you have here.

Reply:
> The first thing you should be checking is whether any of the deleted messages is the current first_message_id of the stream; if none are, then, there's no need to do queries touching the Message table at all.

But let's say in any existing stream the first message sent has been deleted before this implementation. Now after this implementation, if the current first message of that stream gets deleted, this logic will not be able to update the `first_message_id` as it is not equal to the ID of the deleted message. Am I right?

Reply:
Hmm, that's a good point. We should have a comment noting that; I think eventually we could do a migration to eliminate that possibility, but probably we should merge this simpler version that doesn't require it.

Reply:
OK, that sounds good. I have added a test and moved the logic to a separate function.